File: src/com/train/algorithm/search/implementInC++/SudokuSolver.cpp
//
// Created by 罗斯 on 17/9/2022.
//
#include <vector>

using namespace std;

class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        rows_ = vector<vector<int>>(9, vector<int>(10));
        cols_ = vector<vector<int>>(9, vector<int>(10));
        boxes_ = vector<vector<int>>(9, vector<int>(10));

        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                const char c = board[i][j];
                if (c != '.'){
                    int n = c - '0';
                    int bx = j / 3;
                    int by = i / 3;
                    rows_[i][n] = 1;
                    cols_[j][n] = 1;
                    boxes_[by * 3 + bx][n] = 1;
                }
            }
        }
        fill(board,0,0);
    }

private:
    vector<vector<int>> rows_, cols_, boxes_;

    bool fill(vector<vector<char>>& board, int x, int y){
        if (y == 9) return true;

        int nx = (x + 1) % 9;
        int ny = (nx == 0) ? y + 1 : y;

        if (board[y][x] != '.') return fill(board, nx, ny);

        for (int i = 1; i <= 9; ++i) {
            int bx = x / 3;
            int by = y / 3;
            int box_key = by * 3 + bx;
            if (!rows_[y][i] && !cols_[x][i] && !boxes_[box_key][i]) {
                rows_[y][i] = 1;
                cols_[x][i] = 1;
                boxes_[box_key][i] = 1;
                board[y][x] = i + '0';
                if (fill(board, nx, ny)) return true;
                board[y][x] = '.';
                boxes_[box_key][i] = 0;
                cols_[x][i] = 0;
                rows_[y][i] = 0;
            }
        }
        return false;
    }
};

File: src/com/train/algorithm/search/implementInPython/SudokuSolver.py
class Solution(object):
    def solveSudoku(self, board):
        n = 9
        def isValid(r, c, char):
            r, c = int(r), int(c)
            for i in range(9):
                if board[i][c] == char:
                    return False
                if board[r][i] == char:
                    return False
                if board[3 * (r//3)+i//3][3*(c//3)+i%3] == char:
                    return False
            return True

        def solve(r, c):
            if r == n:
                return True
            if c == n:
                return solve(r + 1, 0)
            if board[r][c] == ".":
                for i in range(1, 10):
                    if isValid(r, c, str(i)):
                        board[r][c] = str(i)

                        if solve(r, c + 1):
                            return True
                        else:
                            board[r][c] = "."
                return False
            else:
                return solve(r, c + 1)

        solve(0,0)